internal/backend/auth/authhttpmiddleware/authmiddleware.go

```diff
@@ -0,0 +1,105 @@
+package authhttpmiddleware
+
+import (
+	"net/http"
+	"strings"
+
+	"go.uber.org/fx"
+
+	"go.autokitteh.dev/autokitteh/internal/backend/auth/authcontext"
+	"go.autokitteh.dev/autokitteh/internal/backend/auth/authsessions"
+	"go.autokitteh.dev/autokitteh/internal/backend/auth/authtokens"
+	"go.autokitteh.dev/autokitteh/internal/backend/configset"
+)
+
+type Config struct {
+	Required bool `koanf:"required"`
+}
+
+var Configs = configset.Set[Config]{
+	Default: &Config{Required: true},
+	Dev:     &Config{},
+}
+
+type WrapFunc func(http.Handler) http.Handler
+
+type Deps struct {
+	fx.In
+
+	Cfg      *Config
+	Sessions authsessions.Store `optional:"true"`
+	Tokens   authtokens.Tokens  `optional:"true"`
+}
+
+func newTokensMiddleware(next http.Handler, tokens authtokens.Tokens) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		if user := authcontext.GetAuthnUser(ctx); !user.IsValid() {
+			if auth := r.Header.Get("Authorization"); auth != "" {
+				kind, payload, _ := strings.Cut(auth, " ")
+
+				switch kind {
+				case "Bearer":
+					var err error
+					if user, err = tokens.Parse(payload); err != nil {
+						http.Error(w, "invalid token", http.StatusUnauthorized)
+						return
+					}
+
+					r = r.WithContext(authcontext.SetAuthnUser(ctx, user))
+				default:
+					http.Error(w, "invalid authorization header", http.StatusUnauthorized)
+					return
+				}
+			}
+		}
+
+		next.ServeHTTP(w, r)
+	}
+}
+
+func newSessionsMiddleware(next http.Handler, sessions authsessions.Store) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		if user := authcontext.GetAuthnUser(ctx); !user.IsValid() {
+			session, err := sessions.Get(r)
+			if err != nil {
+				http.Error(w, "invalid session", http.StatusUnauthorized)
+				return
+			}
+
+			if session != nil {
+				r = r.WithContext(authcontext.SetAuthnUser(ctx, session.User))
+			}
+		}
+
+		next.ServeHTTP(w, r)
+	}
+}
+
+func New(deps Deps) WrapFunc {
+	return func(next http.Handler) http.Handler {
+		f := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if deps.Cfg.Required {
+				if user := authcontext.GetAuthnUser(r.Context()); !user.IsValid() {
+					http.Redirect(w, r, "/login", http.StatusTemporaryRedirect)
+					return
+				}
+			}
+
+			next.ServeHTTP(w, r)
+		})
+
+		if deps.Sessions != nil {
```

Comment:
do we still need both sessions and token ?

Reply:
yes, for web ui.